HW0/P5/P5.txt

```diff
@@ -1 +1,108 @@
-# Your answers here.
+1. 
+Bags = 256
+Cashiers = 1
+Time/addition = 1
+>> Steps: cashier adds bag #1 to bag #2, then adds that to bag #3, etc.
+>> Time to add = Bags - 1 = 255 seconds
+
+2.
+Bags = 256
+Cashiers = 8
+Time/addition = 1
+>> Steps: Each cashier receives 256/8 = 32 bags, which take 31 seconds to count. Each cashier then turns to another cashier and sums their totals, which takes 1 second and results in 4 new totals. This is repeated until there is only 1 total left.
+>> Time to add = 34 seconds (= 31 seconds for each cashier's initial calculation + 3 seconds to aggregate everything)
+
+3.
+Bags = 256
+Cashiers = Infinite
+Time/addition = 1
+>> Steps: Same as #2, but in the first step each cashier receives only 2 bags, which take 1 second to add up (ie. there are infinite cashiers, but only 128 are actually used). As in #2, each cashier turns to another cashier and sums their totals; this is repeated until there is only 1 total left.
+>> Time to add = log-2(256) = 8 seconds
+
+4.
+Bags = N
+Cashiers = Infinite
+Time/addition = 1
+>> Steps: Same as #3, with each initial cashier receiving N/2 bags.
+>> Time to add = log-2(N)
+(see P5.png)
+
+5.
+Bags = 256
+Cashiers = Infinite
+Time/addition = 1
+Time/communication = 1
+>> Steps: Same steps as #2. The time of the cashiers' first calculation is the same, but each subsequent aggregation takes 2 seconds (1 second communication + 1 second addition) instead of 1 second.
+>> Time to add = 2*log(256/2)+1 = 15 seconds
+
+6.
+Bags = 256
+Cashiers = Infinite
+Time/addition = 1
+Time/communication = 1
+Time/handing bag(s) = 1
+
+Scenario 1: Each cashier comes to you and takes 1 bag in turn. The cashiers start communicating/adding while the bags are still being handed out.
+Assume the following notations:
+d = distribution of 1 bag (1 second)
+c = communication across cashiers (1 second)
+s = summing of communicated numbers (1 second)
+At each second, the following transitions are possible:
+c -> s i.e. numbers that have been communicated are summed
+any pair of d and/or s (i.e. dd, ds, ss, sd) -> c i.e. numbers are communicated across cashiers
+
+The following pattern develops:
+1		2		3		4		5		6		7		8
+---		---		---		---		---		---		---		—
+d		d		c		s		s		s		s		s
+		d		d		d		c		c		c		c
+				d		d		d		d		d		d
+
+After second 5, a stationary set of operations develops.  Since all bags have to be distributed, at least 256 steps must be taken.  At the 256th second, the following occurs:
+
+256		257		258		259		260
+---		---		---		---		---
+s		c		s		c		s
+c		s		s
+d
+
```

Comment:
Your Qn 6, Scenario 1 solution is much more efficient than mine. I just left the cashiers standing there dumbly with the bags. kudos!